Feat/98 file upload api

manvinderjit opened the pull request:
## Summary

Add file upload api for uploading files to R2 buckets.

## Tests
WIP - Pending

## Additional information

Closes #98

manvinderjit committed:
feat: add file validation zod schemas

manvinderjit committed:
feat: add file get and upload routes

manvinderjit committed:
db: add schema for uploads table

add migration schema for the uploads table to store files.
add table to DBTables constant.

manvinderjit committed:
feat: add file validation schemas and functions

manvinderjit committed:
feat: update file upload route to save file data

add uploadFile method to save file data.
update the api to save the data and leverage validation functions.

manvinderjit committed:
feat: add apis to get files from r2

add a public api to fetch public files from r2.
add a private api to fetch protected files from r2.
add a data function to fetch file info from uploads table.

madcampos commented on api/src/routes/files/data.ts:
```diff
@@ -0,0 +1,45 @@
+export const fetchFileInfo = async (database: D1Database, fileNameAsId: string, fileAccessLevel: FileAccessLevelType = 'public', isAuthenticated = false) => {
+	// Search all files if requesting protected file and authenticated, otherwise only search public files
+	const dbQuery = fileAccessLevel === 'protected' && isAuthenticated ?
+		`SELECT id AS fileName, mimeType, accessLevel
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities. -->
**thought:** ID and file name should be different if we want to retain the file name.

The ID has to be unique and is use to reference the file. The file name can be anything, and may even be the same name from other files.

Imagine I have `profile-picture.jpg` and you also have a file with the same name. They are two diferent files, but the name is the same.

manvinderjit replied:
Do we need the filename? The id is a uuid being generated.

madcampos replied:
I don't know, it may not be useful for user avatars, but we have to think:  is there a scneario where it will be useful?

manvinderjit replied:
Removed access level entirely because the philosophy of the project is design to be public. The avatar URL is now stored in the `Profile` table.

madcampos commented on api/src/routes/files/data.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const fetchFileInfo = async (database: D1Database, fileNameAsId: string, fileAccessLevel: FileAccessLevelType = 'public', isAuthenticated = false) => {
+	// Search all files if requesting protected file and authenticated, otherwise only search public files
+	const dbQuery = fileAccessLevel === 'protected' && isAuthenticated ?
+		`SELECT id AS fileName, mimeType, accessLevel
+			FROM ${DBTables.UPLOADS}
+			WHERE
+				id = ?
+			LIMIT 1
+			` :
+		`SELECT id AS fileName, mimeType, accessLevel
+			FROM ${DBTables.UPLOADS}
+			WHERE
+				id = ?
+				AND accessLevel = 'public'
+			LIMIT 1`;
```
```suggestion
	const accessLevelCheck = fileAccessLevel === 'protected' && isAuthenticated ?  "AND accessLevel = 'public'" : '';
	const dbQuery = `
		SELECT id AS fileName, mimeType, accessLevel
		FROM ${DBTables.UPLOADS}
		WHERE
			id = ?
			${accessLevelCheck}
		LIMIT 1
	`;
```

madcampos commented on api/src/routes/files/data.ts:
```diff
@@ -0,0 +1,45 @@
+	const now = new Date().toISOString();
+
+	const { success } = await database.prepare(`
+		INSERT INTO ${DBTables.UPLOADS} (id, schemaVersion, profileId, type,  insertedAt, mimeType, accessLevel) VALUES (?, ?, ?, ?, ?, ?, ?)
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** What happens when the user uploads a file twice? Will it overwrite the previous one?

manvinderjit replied:
Each file is treated independent. So, the new file will have a separate uuid and will be stored accordingly. 

1. For files that have a specific purpose, like `avatar`, we can get the latest one by `insertedAt`. Alternatively, We have an `avatar` field in the `profile` table. We can set its value to the `uuid` of the image that is selected or latest. 
And we can also show a user their avatar history, if such a feature is planned. 
2. For files that have no specific purpose, we have each file existing separately.
3. For removing old files, we can use a worker to clean the files that are greater than x days and remove them from database.

madcampos replied:
Avatar history is not planned. I don't think any place does that.

A blanket cleanup of files is not an ideal solution, imagine a user uploads their avatar when creating an account and never changes it. Two years have passed. Should this file be removed or kept?

An alternative would be to upload multiple files and then remove the previous one, but how would you identify what file is "the previous one"?

manvinderjit replied:
Resolve: New file overwrites old file URL in database. Existing upload is deleted from R2 after new one is uploaded.

madcampos commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+	}),
+	async (context) => {
+		try {
+			const { filename } = context.req.valid('param');
+
+			if (!filename) {
+				return context.json({ message: 'Invalid or no filename provided!' }, StatusCodes.BAD_REQUEST);
+			}
```
Zod should already check for missing parameters

manvinderjit replied:
Added this as a double check in a rare case if zod falters or someone changes the zod schema. I can remove if it is undesired.

madcampos replied:
Yeah, better remove it.

kbventures replied:
@manvinderjit, we're not introducing code to plan against Zod failing. As for the second scenario, we just need to make sure people don't remove Zod schemas.

madcampos commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+		}
+
+		const fileMimeType = file.type as AllowedFileMimeType;
+
+		// Validate file extension
+		if (!validateFileExtension(file.name, fileMimeType)) {
+			return context.json({ message: 'Invalid image file extension. Only JPEG and PNG are allowed!' } satisfies StatusResponse, StatusCodes.BAD_REQUEST);
+		}
```
This check should be before the public/private one

manvinderjit replied:
Actually the public/private check should be just before try/catch but your suggestion stands.

madcampos commented on api/src/routes/files/validation.ts:
```diff
@@ -0,0 +1,72 @@
+
+export type FileNameParam = z.infer<typeof FileNameParamSchema>;
+
+export const AllowedFileMimeTypesOptions = ['image/png', 'image/jpeg', 'image/jpg'] as const;
```
```suggestion
export const AllowedFileMimeTypesOptions = ['image/png', 'image/jpeg'] as const;
```

The mime type is `image/jpeg`, it applies to two extensions: `jpg` and `jpeg`

manvinderjit replied:
Thank you for sharing, Good to know. I thought they were separate.

madcampos commented on api/src/routes/files/validation.ts:
```diff
@@ -0,0 +1,72 @@
+export type AllowedFileMimeType = z.infer<typeof AllowedFileMimeTypeSchema>;
+
+export const mimeToExtensionsMap: Record<AllowedFileMimeType, string[]> = {
+	'image/jpeg': ['jpg', 'jpeg'],
+	'image/jpg': ['jpg', 'jpeg'],
+	'image/png': ['png']
+};
```
I think it is better to use the [`mime`](https://www.npmjs.com/package/mime) library for this.

manvinderjit replied:
Do we really want to add a whole library for this? The map is just to check the MimeType and the file extension are the same. It is not being used for anything else. i could have used a switch statement but thought this will be easier.

madcampos replied:
For a small subset of extensions and mimes, it is okay to keep a manual list. As long as they are correctly represented, please read those two links:
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/MIME_types
- https://www.iana.org/assignments/media-types/media-types.xhtml

Also note that a mime type can correspond to [multiple extensions](https://en.wikipedia.org/wiki/JPEG#:~:text=.jpg%2C%20.jpeg%2C%20.jpe%0A.jif%2C%20.jfif%2C%20.jfi).

kbventures replied:
@manvinderjit, the lite version is extremely small and is widely used, so it would save you from doing research. 

<img width="810" height="202" alt="image" src="https://github.com/user-attachments/assets/5df7e1c5-e712-4237-b208-1cbed96b819d" />

manvinderjit replied:
Updated the code.
Right now we only have two MIME types. We can add a library in future if we expand them.

madcampos commented on api/src/routes/files/validation.ts:
```diff
@@ -0,0 +1,72 @@
+}
+
+export const validateFileExtension = (fileName: string, mimeType: string): boolean => {
+	const lastDotIndex = fileName.lastIndexOf('.');
+	const lastSlashIndex = Math.max(fileName.lastIndexOf('/'), fileName.lastIndexOf('\\'));
+
+	// No extension or extension is at the end (e.g., "file.")
+	if (lastDotIndex <= lastSlashIndex || lastDotIndex === fileName.length - 1) {
+		return false;
+	}
```
This can be simplified to:
```javascript
testString.split('.').pop();
```

It will give you the file extension. If it is an empty string, it means the name ends with a `.`, if it is anything other than letters and numbers (Regex: `/[a-z0-9]+/iu`), it is not an extension.

manvinderjit replied:
As we are dealing with R2 prefixes that will have delimiter `/`, the suggested code will incorrectly treat  a scenario like `"/path.to/file"` as "to/file" and having extension "file", because split('.') doesn't determine if the dot was in the prefix or the file name.
While this may rarely occur, but having this edge case covered is recommended.

manvinderjit committed:
fix: cleanup db query

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: remove redundant mime type

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: resolve linting and formatting errors

manvinderjit committed:
fix: correct validation hierarchy

move file access level check after validating file extension

manvinderjit committed:
fix: reverse accessLevelCheck logic

reverse the logic in accessLevelCheck for files.

manvinderjit committed:
feat: delete orphaned file from r2

delete orphaned file from r2 if file data not stored in database.

manvinderjit committed:
fix: trim filename received as param

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+	}),
+	async (context) => {
+		try {
+			const filename = context.req.valid('param').filename?.trim();
```
Trim should definitely be in the validation file, and the filename existence as well. 

Something like:

```
const filenameSchema = z.string().trim().min(1, "Filename cannot be empty");
```

manvinderjit replied:
Removed filename altogether because now the api is refactored to fetch files on the R2 prefix and not just the filename.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+		try {
+			const filename = context.req.valid('param').filename?.trim();
+
+			if (!filename) {
```
Does this still need to exist if we action the previous comment?

manvinderjit replied:
Removed, same as above.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+				return context.json({ message: 'Profile not found' } satisfies StatusResponse, StatusCodes.NOT_FOUND);
+			}
+
+			const filename = context.req.valid('param').filename?.trim();
```
These should be in the Zod validation file as stated above.

manvinderjit replied:
Removed.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+			const fileInfo = await fetchFileInfo(context.env.Database, filename, 'protected', true);
+
+			if (fileInfo) {
+				// TODO: Add a check to verify if the user is allowed to access the file.
```
This sounds like something that shouldn't be optional, and Git issue should be created and shown below the comment.

manvinderjit replied:
Removed access levels all together for the moment.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+				const r2Bucket = context.env.ExportedFiles;
+				const file = await r2Bucket.get(filename);
+
+				// If file is found, return it
```
This is self-explanatory. Comment is unnecessary.

kbventures commented on api/src/routes/files/data.ts:
```diff
@@ -0,0 +1,41 @@
+	const { success } = await database.prepare(`
+		INSERT INTO ${DBTables.UPLOADS} (id, schemaVersion, profileId, type,  insertedAt, mimeType, accessLevel) VALUES (?, ?, ?, ?, ?, ?, ?)
+	`)
+		.bind(fileNameAsId, SCHEMA_VERSION, profileId, fileType, now, fileMimeType, accessLevel)
```
now?

This needs to be renamed.

manvinderjit replied:
Outdated code. No longer required.

kbventures commented on api/src/routes/files/data.ts:
```diff
@@ -0,0 +1,41 @@
+import { DBTables, SCHEMA_VERSION } from '../../utils/db.ts';
+import type { AllowedFileMimeType, FileAccessLevelType, UploadedFileType } from './validation.ts';
+
+export const fetchFileInfo = async (database: D1Database, fileNameAsId: string, fileAccessLevel: FileAccessLevelType = 'public', isAuthenticated = false) => {
```
Question:

It seems like we're mixing authentication concerns at the data layer. 

Why is this not handled by our middleware?

I realize this might force us to end up with two functions,  but their concerns will then be unambiguous.

manvinderjit replied:
Refactored code. No longer required.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+				const r2Bucket = context.env.ExportedFiles;
+				const file = await r2Bucket.get(filename);
+
+				// If file is found, return it
```
Most of this code is self-explanatory.  No needs for notes for this.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+				}
+			}
+
+			// If the file is not found
```
Same as note above.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+				return context.json({ message: 'Invalid or no filename provided!' }, StatusCodes.BAD_REQUEST);
+			}
+
+			// Fetch file data to determine if it exists
```
No need comments here.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+		const file = formData.get('file') as File;
+		const fileType = formData.get('type') as UploadedFileType;
+
+		// Check if file was uploaded by the user
```
Remove this comment. Same reason as above.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+
+		const fileMimeType = file.type as AllowedFileMimeType;
+
+		// Validate file extension
```
Unnecessary comments.

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,270 @@
+			const r2Bucket = context.env.ExportedFiles;
+			const r2UploadResult = await r2Bucket.put(fileName, stream);
+
+			// TODO: Determine if we need to create thumbnails
```
Create Github issue and add the link below.

manvinderjit replied:
Done
https://github.com/torontojs/communityhub/issues/270

kbventures commented on api/src/routes/files/validation.ts:
```diff
@@ -0,0 +1,71 @@
+
+	const extension = fileName.slice(lastDotIndex + 1).toLowerCase();
+
+	// Return false if MIME type isn't allowed
```
No need for comments here.

kbventures reviewed (changes requested):
Great work.

manvinderjit committed:
fix: remove unwanted comments

remove unwanted comments from files/index.ts

manvinderjit committed:
fix: remove unwanted comments

remove unwanted comments from files/validation.ts

manvinderjit committed:
db: rename and update uploads schema migration

rename the file schema migration from user-uploads to create-uploads.
add a filename field to keep the filename separate from id.
remove the accessLevel field, enabling only public types.

manvinderjit committed:
refactor: store avatar url in profile table directly

update avatar url in the profile table instead of creating a separate uploads table.
delete uploads table.

manvinderjit committed:
cleanup: access based file validation methods

clean up file validation methods based on public and private access.

manvinderjit committed:
refactor: file apis to use profile table for avatar

update the POST files/upload api to upload avatar in Profile table.
update the GET files/ api to allow public file access from R2 bucket using file prefix.

manvinderjit committed:
Merge latest changes from 'origin/dev' into feat/98-file-upload-api

manvinderjit committed:
fix: remove UPLOADS from DBTables const

kbventures commented on api/src/routes/files/index.ts:
```diff
@@ -0,0 +1,174 @@
+	defaultHook: statusResponseFormatter
+});
+
+// NOTE: Add more specific GET routes before the following route only
```
Love to see this new cool feature. 

Some comments. 

A public service for a resource which cost money?  That sounds like  a bad idea for whomevers credit card goes into the Cloudflare account.  The service needs to be scoped
 to retreving Avatars ONLY and with authentication.

kbventures replied:
We're building a image uplaod service specificaly not a file upload service.  I think the naming should reflect that and has mentioned above scoped for that specifially.

kbventures replied:
As per our conversation, we need to code a check to ensure we don't go voer the "free" tier limit.  There's no way to to have alarms or limits setup at the admin console level.   Checking everything we utilize(resource) is not optimal but we can find a more optimal way after the MVP is in production.

bcha92 replied:
+1 on the limiter. If we are using a tier-limited service, we would need to manually setup a logger that tracks this and send an alert once it reaches a certain threshold.

kbventures reviewed (changes requested):
Please include at least one hurl interagration tests and create a github issue to create unit tests for this feature(image upload).  Our unit tests setup is almost ready.
